Add Config validation to build stage to prevent runtime panics

When creating a custom zap.Config there seems to be only minimal validation of the configuration that results in runtime panics if certain properties are missing. This is dangerous as all required configuration properties should be checked at the build stage. 

There is probably a reason this has not been done, but opening an issue nonetheless.

## Example

```
	cfg := zap.Config{
		Encoding:    "json",
		OutputPaths: []string{"stderr"},
		EncoderConfig: zapcore.EncoderConfig{
			MessageKey: "msg",
		},
	}

	logger, err := cfg.Build()
	if err != nil {
		panic(err)
	}

	logger.Sugar().Errorw("Problem here", "foo", "bar") // Panic
```

This results in the following panic:

```

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x12696c6]

goroutine 1 [running]:
go.uber.org/atomic.(*Int32).Load(...)
	/Users/owner/go/pkg/mod/go.uber.org/atomic@v1.5.1/atomic.go:41
go.uber.org/zap.AtomicLevel.Level(...)
	/Users/owner/go/pkg/mod/go.uber.org/zap@v1.13.0/level.go:97
go.uber.org/zap.AtomicLevel.Enabled(0x0, 0x10b5c02, 0xc000098c30)
	/Users/owner/go/pkg/mod/go.uber.org/zap@v1.13.0/level.go:92 +0x6
go.uber.org/zap.(*SugaredLogger).log(0xc000055f00, 0xc000010002, 0x1312a3e, 0xc, 0x0, 0x0, 0x0, 0xc000055f30, 0x2, 0x2)
	/Users/owner/go/pkg/mod/go.uber.org/zap@v1.13.0/sugar.go:221 +0x1aa
go.uber.org/zap.(*SugaredLogger).Errorw(...)
	/Users/owner/go/pkg/mod/go.uber.org/zap@v1.13.0/sugar.go:191
main.main()
	/Users/owner/zaplog/zaplog.go:22 +0x27b
exit status 2
```

The issue here is we are missing a "Level"

```
Level:            zap.NewAtomicLevelAt(zapcore.DebugLevel),
```

However, there are other issues as well. Suppose you specify the "TimeKey", but not the "EncodeTime".

```
	cfg := zap.Config{
		Level:       zap.NewAtomicLevelAt(zapcore.DebugLevel),
		Encoding:    "json",
		OutputPaths: []string{"stderr"},
		EncoderConfig: zapcore.EncoderConfig{
			MessageKey: "msg",
			TimeKey:    "ts",
		},
	}
```

This configuration will build without error, but you again will be faced with `runtime error: invalid memory address or nil pointer dereference` on use.

--- a/config.go
+++ b/config.go
@@ -174,6 +174,13 @@ func (cfg Config) Build(opts ...Option) (*Logger, error) {
 		return nil, err
 	}
 
+	if cfg.Level == (AtomicLevel{}) {
+		cfg.Level = NewAtomicLevelAt(InfoLevel)
+		if cfg.Development {
+			cfg.Level = NewAtomicLevelAt(DebugLevel)
+		}
+	}
+
 	log := New(
 		zapcore.NewCore(enc, sink, cfg.Level),
 		cfg.buildOptions(errSink)...,
@@ -239,5 +246,12 @@ func (cfg Config) openSinks() (zapcore.WriteSyncer, zapcore.WriteSyncer, error) 
 }
 
 func (cfg Config) buildEncoder() (zapcore.Encoder, error) {
+	if cfg.EncoderConfig.TimeKey != "" && cfg.EncoderConfig.EncodeTime == nil {
+		cfg.EncoderConfig.EncodeTime = zapcore.EpochTimeEncoder
+		if cfg.Development {
+			cfg.EncoderConfig.EncodeTime = zapcore.ISO8601TimeEncoder
+		}
+	}
+
 	return newEncoder(cfg.Encoding, cfg.EncoderConfig)
 }

--- a/config.go
+++ b/config.go
@@ -244,9 +244,5 @@ func (cfg Config) openSinks() (zapcore.WriteSyncer, zapcore.WriteSyncer, error) 
 }
 
 func (cfg Config) buildEncoder() (zapcore.Encoder, error) {
-	if cfg.EncoderConfig.TimeKey != "" && cfg.EncoderConfig.EncodeTime == nil {
-		return nil, fmt.Errorf("missing EncodeTime in EncoderConfig")
-	}
-
 	return newEncoder(cfg.Encoding, cfg.EncoderConfig)
 }

--- a/config.go
+++ b/config.go
@@ -21,6 +21,7 @@
 package zap
 
 import (
+	"fmt"
 	"sort"
 	"time"
 
@@ -175,10 +176,7 @@ func (cfg Config) Build(opts ...Option) (*Logger, error) {
 	}
 
 	if cfg.Level == (AtomicLevel{}) {
-		cfg.Level = NewAtomicLevelAt(InfoLevel)
-		if cfg.Development {
-			cfg.Level = NewAtomicLevelAt(DebugLevel)
-		}
+		return nil, fmt.Errorf("missing Level")
 	}
 
 	log := New(
@@ -247,10 +245,7 @@ func (cfg Config) openSinks() (zapcore.WriteSyncer, zapcore.WriteSyncer, error) 
 
 func (cfg Config) buildEncoder() (zapcore.Encoder, error) {
 	if cfg.EncoderConfig.TimeKey != "" && cfg.EncoderConfig.EncodeTime == nil {
-		cfg.EncoderConfig.EncodeTime = zapcore.EpochTimeEncoder
-		if cfg.Development {
-			cfg.EncoderConfig.EncodeTime = zapcore.ISO8601TimeEncoder
-		}
+		return nil, fmt.Errorf("missing EncodeTime in EncoderConfig")
 	}
 
 	return newEncoder(cfg.Encoding, cfg.EncoderConfig)

--- a/config.go
+++ b/config.go
@@ -175,7 +175,7 @@ func (cfg Config) Build(opts ...Option) (*Logger, error) {
 		return nil, err
 	}
 
-	if cfg.Level == AtomicLevel{} {
+	if cfg.Level == (AtomicLevel{}) {
 		return nil, fmt.Errorf("missing Level")
 	}
 

--- a/config_test.go
+++ b/config_test.go
@@ -27,6 +27,7 @@ import (
 
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
+	"go.uber.org/zap/zapcore"
 )
 
 func TestConfig(t *testing.T) {
@@ -103,6 +104,42 @@ func TestConfigWithInvalidPaths(t *testing.T) {
 			cfg.ErrorOutputPaths = []string{tt.errOutput}
 			_, err := cfg.Build()
 			assert.Error(t, err, "Expected an error opening a non-existent directory.")
+		})
+	}
+}
+
+func TestConfigWithMissingAttributes(t *testing.T) {
+	tests := []struct {
+		desc      string
+		cfg       Config
+		expectErr string
+	}{
+		{
+			desc: "missing_level",
+			cfg: Config{
+				Encoding: "json",
+			},
+			expectErr: "missing Level",
+		},
+		{
+			desc: "missing_encoder_time_in_encoder_config",
+			cfg: Config{
+				Level:    NewAtomicLevelAt(zapcore.InfoLevel),
+				Encoding: "json",
+				EncoderConfig: zapcore.EncoderConfig{
+					MessageKey: "msg",
+					TimeKey:    "ts",
+				},
+			},
+			expectErr: "missing EncodeTime in EncoderConfig",
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.desc, func(t *testing.T) {
+			cfg := tt.cfg
+			_, err := cfg.Build()
+			assert.Equal(t, err.Error(), tt.expectErr)
 		})
 	}
 }

--- a/config_test.go
+++ b/config_test.go
@@ -115,14 +115,14 @@ func TestConfigWithMissingAttributes(t *testing.T) {
 		expectErr string
 	}{
 		{
-			desc: "missing_level",
+			desc: "missing level",
 			cfg: Config{
 				Encoding: "json",
 			},
 			expectErr: "missing Level",
 		},
 		{
-			desc: "missing_encoder_time_in_encoder_config",
+			desc: "missing encoder time in encoder config",
 			cfg: Config{
 				Level:    NewAtomicLevelAt(zapcore.InfoLevel),
 				Encoding: "json",

--- a/config_test.go
+++ b/config_test.go
@@ -139,7 +139,7 @@ func TestConfigWithMissingAttributes(t *testing.T) {
 		t.Run(tt.desc, func(t *testing.T) {
 			cfg := tt.cfg
 			_, err := cfg.Build()
-			assert.Equal(t, err.Error(), tt.expectErr)
+			assert.Equal(t, tt.expectErr, err.Error())
 		})
 	}
 }

--- a/config_test.go
+++ b/config_test.go
@@ -139,6 +139,7 @@ func TestConfigWithMissingAttributes(t *testing.T) {
 		t.Run(tt.desc, func(t *testing.T) {
 			cfg := tt.cfg
 			_, err := cfg.Build()
+			require.Error(t, err)
 			assert.Equal(t, tt.expectErr, err.Error())
 		})
 	}

--- a/encoder.go
+++ b/encoder.go
@@ -62,6 +62,10 @@ func RegisterEncoder(name string, constructor func(zapcore.EncoderConfig) (zapco
 }
 
 func newEncoder(name string, encoderConfig zapcore.EncoderConfig) (zapcore.Encoder, error) {
+	if encoderConfig.TimeKey != "" && encoderConfig.EncodeTime == nil {
+		return nil, fmt.Errorf("missing EncodeTime in EncoderConfig")
+	}
+
 	_encoderMutex.RLock()
 	defer _encoderMutex.RUnlock()
 	if name == "" {

